### ISSUE TITLE
`blur` function is too slow

### ISSUE THREAD

**Issue by Aloxaf** [image-rs/image#986]
`blur` is too slow, `imageproc::filter::gaussian_blur_f32` is faster than it, but still slow.

## Reproduction steps

```Rust
use image::imageops::blur;
use image::{Rgb, RgbImage};
use imageproc::drawing::draw_filled_circle_mut;
use imageproc::filter::gaussian_blur_f32;
use std::time::Instant;

fn main() {
    let mut image = RgbImage::new(1000, 1000);
    draw_filled_circle_mut(&mut image, (500, 500), 500, Rgb([255, 255, 255]));

    let start = Instant::now();
    blur(&image, 100.0).save("rust_1.png").unwrap();
    dbg!(start.elapsed());

    let start = Instant::now();
    gaussian_blur_f32(&image, 100.0).save("rust_2.png").unwrap();
    dbg!(start.elapsed());
}
```

result is 
```
[src/main.rs:13] start.elapsed() = 12.362780732s
[src/main.rs:17] start.elapsed() = 3.335411994s
```

and Pillow
```python
from PIL import Image, ImageDraw, ImageFilter
from time import time

img  = Image.new('RGB', (1000, 1000))
draw = ImageDraw.Draw(img)
draw.ellipse((0, 0, 1000, 1000), fill='#ffffff')

start = time()
img.filter(ImageFilter.GaussianBlur(100)).save('python.png')
print(f'{time() - start}s')
```

result is `0.1433885097503662s`

Here is the final image 

rust_1.png (it's a little strange...
![rust_1](https://user-images.githubusercontent.com/17017672/60798873-994a4280-a1a4-11e9-8db5-78528b2605e1.png)

rust_2.png
![rust_2](https://user-images.githubusercontent.com/17017672/60798882-9e0ef680-a1a4-11e9-9c93-007ca0253153.png)

python.png
![python](https://user-images.githubusercontent.com/17017672/60798894-a36c4100-a1a4-11e9-99c2-79e41f4e87aa.png)

**Comment by HeroicKatora** [comment]
Just to clarify, PIL uses a C version indirectly https://github.com/python-pillow/Pillow/blob/ab9a25d623fdd7f8de3e724b538f5660eac589ae/src/libImaging/BoxBlur.c#L294

It's still somewhat embarrasingly slow.

**Comment by Aloxaf** [comment]
@HeroicKatora Yes，maybe porting it to rust will be a good idea？

**Comment by theotherphil** [comment]
Pillow appears to use (a variant on) iterated box filtering, as defined in http://www.mia.uni-saarland.de/Publications/gwosdek-ssvm11.pdf

There's an open imageproc ticket to implement this: https://github.com/image-rs/imageproc/issues/93

**Comment by theotherphil** [comment]
I'll have a look at implementing this over the coming weekend.

**Comment by Shnatsel** [comment]
There is actually a linear-complexity blur already implemented in Rust, where execution time is independent of blur radius: https://github.com/fschutt/fastblur

A copy of it was copied into resvg codebase and polished to match Chrome blur. You can find it in [this file](https://github.com/RazrFalcon/resvg/blob/master/src/filter.rs), search for `box_blur`. It should be fairly easy to extract it back into a common crate and/or copy it into `image`.

**Comment by vadixidav** [comment]
You may want to take a look at the code [here](https://github.com/rust-cv/akaze/blob/cc8d1b1f06b8b518ced5152251d80b1a64060344/src/fed_tau.rs). It was originally based on a different paper, written by the author of AKAZE, ported to rust by someone else, and then modified later by me (so it has been touched by many hands). It makes use of the following papers:

* S. Grewenig, J. Weickert, C. Schroers, A. Bruhn. Cyclic Schemes for PDE-Based Image Analysis. Technical Report No. 327, Department of Mathematics, Saarland University, Saarbrücken, Germany, March 2013
* S. Grewenig, J. Weickert, A. Bruhn. From box filtering to fast explicit diffusion. DAGM, 2010

It is actually not used for guassian blur in akaze itself. I am also not 100% sure it can help speed up guassian blur, but I think it can based on what I have seen in its use in akaze. I am no expert. The way this algorithm works is that it determines a number of diffusion steps that start small and grow bigger as the stability increases (which may not apply to guassian blur?). See [this file](https://github.com/rust-cv/akaze/blob/cc8d1b1f06b8b518ced5152251d80b1a64060344/src/nonlinear_diffusion.rs) for how the specialized blur is being applied (with ndarray). I was able to get pretty good speed by writing my filters with ndarray like that. My situation is a bit more specific to this library, but the code is free for you to take.

**Comment by RReverser** [comment]
Also noticed this while comparing Wasm version of `image::imageops::blur` and JS implementation of Gaussian blur from https://github.com/nodeca/glur.

At high radius values the difference becomes ridiculous - e.g. for 3872x2592 image and radius 300px the image crate's version executes in 94 seconds, while JS executes in 0.6 seconds (same as for any other radius).

Having linear implementation in image crate would help a lot. `fastblur` from @Shnatsel's link looks promising, but it would be nice to integrate it into `image` or `imageproc` instead of relying on a Git dependency that is somewhat harder to discover.

**Comment by torfmaster** [comment]
If anyone is interested: I created a first draft to resolve this issue in https://github.com/image-rs/image/pull/2302 and I am willing to finalize this.

**Comment by Shnatsel** [comment]
FWIW a new crate with multiple fast blur implementations was just released: https://github.com/awxkee/libblur

It relies on unsafe SIMD intrinsics, so it might not be usable as-is for `image` due to the no-unsafe policy, but we may be able to uplift parts of it anyway.

**Comment by awxkee** [comment]
> FWIW a new crate with multiple fast blur implementations was just released: https://github.com/awxkee/libblur
> 
> It relies on unsafe SIMD intrinsics, so it might not be usable as-is for `image` due to the no-unsafe policy, but we may be able to uplift parts of it or perhaps reuse the fallback implementations.

I can extract part of gaussian just for `image` crate with `forbid unsafe` if that makes sense.
I reworked convolution part without unsafe and situation looks something like that for RGB image 5000x4000:

```
fast_blur sigma 5: 793.836375ms
pure gaussian_blur 5 kernel size: 30.273625ms

fast_blur sigma 15: 803.334958ms
pure gaussian_blur 15 kernel size: 60.557ms

fast_blur sigma 35: 807.447583ms
pure gaussian_blur 35 kernel size: 133.914708ms

fast_blur sigma 151: 811.069792ms
pure gaussian_blur 151 kernel size: 495.628208ms

fast_blur sigma 251: 847.268125ms
pure gaussian_blur 151 kernel size: 875.879ms

pure gaussian_blur 251 kernel size with rayon: 125.079042ms
```

Pure gaussian blur parallelization works pretty fine so adding `rayon` instantly adds x10 performance, and multi-threading for blur's are preferred.

Perhaps I can also add stack blur that have O(1) in big O notation, which will be about ~x5-10 faster than blur based on [CLT](https://en.wikipedia.org/wiki/Central_limit_theorem), which used here on `fast_blur` as far as I can tell. 
Also blur's based on CLT have pretty high convergence with low controlling level, if you noticed that, and images became completely blurred sharply.

Here is stackblur time with disabled SIMD and multi-threading for the same image.

```
stackblur radius 151: 168.549ms
```

Even with `forbid unsafe` I wouldn't expect that it'll slowdown 8 times.

**Comment by Shnatsel** [comment]
Image's `fast_blur` was contributed very recently, and has not received a great deal of attention yet. I'd be happy to see a PR migrating it to a faster algorithm.

I am somewhat concerned about the growing implementation complexity, but I think we could take it on for common and performance-sensitive operations such as resizing and blurring.

**Comment by awxkee** [comment]
Everything of this actually sounds that it is pretty unreasonable :)

I do not think that blur is so common, might be spreaded a bit, but common I think this is not.

And for PR with this pixel store and organization and accessing them in the crate, it is the thing I'd like definitely stay away of that, and size of required code is not small, so complexity will grow noticable.

**Comment by Shnatsel** [comment]
Blur operation isn't that common but it usually does become a bottleneck whenever it crops up, so I think optimizing it is still worthwhile.

**Comment by awxkee** [comment]
May agree.

But issue with pixels storage, lack of really random access with iterator, lack of generic traits still an issue.

I may publish separate crate, but dealing with native problems of crate makes it immediately unreasonable.

**Comment by Shnatsel** [comment]
If you could list the exact problems with the crate's API that prevent this kind of algorithms from being written on top of it in https://github.com/image-rs/image/issues/2358, that would be very helpful!

We're looking to improve the API and ship v1.0 sometime in the foreseeable future, so info on what exactly is missing right now would be very helpful.

**Comment by awxkee** [comment]
Raw row data must be fully accessible by each color component seperately or by set for ex. full RGBA ( this is optional, if each color component accesible this is completely fine ), with iterator without any checks and this call must be `[inline(always)]`.

This can be done by continious allocated vector which is pretty common and very nice.

Or by `brows`, if necessary, it is very good sometimes, and sometimes are not, however, might confusing some who are not familiar with it. If y're not familiar it is an approach where instead of contiguous memory you can do gaps in layout and stores only reference to start ot the row, so your real image is a slice of row references ```&[&[T]]```. 

This at least will make performant methods like [this one](https://github.com/image-rs/image/blob/25fbd3f24a1f79035d031ef7f9391b3f7ccffb49/src/imageops/affine.rs#L91) and actually will prevent some people of doing bad things because someone, who starts, will search how to do so and will see this one.

For any real algorithms all generic traits must be implemented also. And `all` is a very vague definition because sometimes you'll need to make bitxor, or shift right, sometimes use fma when devices have it, and sometimes take a natural logarithm or get `π` from trait.

At the very least, all basic arithmetic must conform.

**Comment by awxkee** [comment]
The thing is that for convolution for example y'll have to access each pixel many, many, many times, for example for blurring with kernel size = 151, on each pixel in the image y'll have to access at least of 150+150 pixels around at very least, if you can't optimize this is out, this is the end.

**Comment by Shnatsel** [comment]
Yes, `GenericImageView` not allowing viewing rows is a known issue: #2300

**Comment by Shnatsel** [comment]
Thank you for the input! I've linked your message from the relevant issues on the bug tracker. We'll see what we can do about these limitations.

**Comment by awxkee** [comment]
Yup, if this reasonable, I still can do blur in separate crate.

Because all these limits can take years :)

**Comment by Shnatsel** [comment]
We're hoping to start a concerted effort on fixing the API sometime around January. But we won't know if they're any good until someone builds a complex image processing algorithm on top of them. Would you be available then to kick the tires of the new APIs and try to port stack blur to them?

**Comment by awxkee** [comment]
Yes, sure, when API looks like at least it worth trying, then I can try to do something on top of this.
I wouldn't say stack blur is complex, but ok :)

Let me know if I can help.

**Comment by Shnatsel** [comment]
https://github.com/image-rs/image/pull/2388 attempted to tackle this but we didn't have the code review capacity at the time. It should be revisited.

**Comment by awxkee** [comment]
I figured out a better way to implement the filter — that’s why I closed the PR.

However, there were also some fundamental issues that I’d like to understand how they are expected to be handled in advance.

Ideal sigma size for kernel size is `0.3 * ((kernel_size - 1.) * 0.5 - 1.) + 0.8` whereas implementation in `image` currently assumes `kernel_size * 2 + 1 = sigma` that is obviously wrong. That makes kernel overly flattened with obvious aliasing, you can see an example in PR.

I don't see too much sense backing incorrect API with correct implementation. So this likely requires a breaking change to make a correct API.

Finally, since this would involve breaking changes, I don’t want to turn a few hours fix into a PR that rots for an year or half just because semver prevents from addressing it properly.

**Comment by Shnatsel** [comment]
It's okay to break the API. The current main already made breaking changes compared to the 0.25.x series. So the time for semver-breaking changes is now!

**Comment by awxkee** [comment]
Let's see what I can do.

Also some implementation questions:
Do you want fixed point implementations for `u8` or `u16` instead of `f32`? Or just pure `f32` convolution is needed?
Do you want a special smoothing cases for kernel size 3/5/7 ?
If special cases is wanted then do you expect them to be bit exact?

Please, see `fn gaussian_blur_indirect<I: GenericImageView>` in previous PR that's the only way I find out how to handle GenericImageView. If I missed something and there’s a better approach, please let me know.

**Comment by Shnatsel** [comment]
`GenericImageView` is probably too restrictive for an efficient convolution implementation. There are several outstanding issues to make it less general and allow at least viewing rows as contiguous slices. It isn't implemented yet.

**Comment by awxkee** [comment]
`fn gaussian_blur_indirect<I: GenericImageView>` just converts `GenericImageView` to raw buffer. It was done to avoiding API removal from trait. In fact since now breaking is allowed I could remove it, but it's better to understand what is expected.

**Comment by awxkee** [comment]
Here is attempt n2 #2496.

Since the questions [here](https://github.com/image-rs/image/issues/986#issuecomment-2983921138) weren't answered, everything was done based on my own preferences.
